sam/bamRead_test.go

```diff
@@ -0,0 +1,202 @@
+	bgSam "github.com/biogo/hts/sam"
+	"github.com/vertgenlab/gonomics/cigar"
+	"github.com/vertgenlab/gonomics/dna"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"testing"
+)
+
+const bamTestfile string = "../bgzf/testdata/test.bam"
+const samTestfile string = "../bgzf/testdata/test.sam"
+
+func TestReadBam(t *testing.T) {
+	var err error
+	r, header := OpenBam(bamTestfile)
+	expected, expHeader := Read(samTestfile)
+
+	// -1 for samtools view line that made the sam file
+	if len(header.Text) != len(expHeader.Text)-1 {
+		t.Errorf("problem reading bam")
+	}
+
+	for i := range header.Text {
+		if header.Text[i] != expHeader.Text[i] {
+			t.Errorf("problem reading bam")
+		}
+	}
+
+	if len(header.Chroms) != len(r.refs) {
+		t.Errorf("plain text and bam header chroms do not match")
+	}
+
+	for i := range r.refs {
+		if r.refs[i] != header.Chroms[i] {
+			t.Errorf("plain text and bam header chroms do not match")
+		}
+	}
+
+	var actual []Sam
+	for {
+		var newSam Sam
+		_, err = DecodeBam(r, &newSam)
+		if err == io.EOF {
+			break
+		}
+		actual = append(actual, newSam)
+	}
+
+	if !equalExceptExtra(actual, expected) {
+		t.Errorf("problem reading bam")
+	}
+
+	err = r.Close()
+	if err != nil {
+		t.Errorf(err.Error())
+	}
+
+	if len(actual) != len(expected) {
+		t.Error("problem reading bam")
+	}
+
+	for i := range actual {
+		if actual[i].String() != expected[i].String() {
+			t.Error("problem reading bam")
+		}
+	}
+}
+
+func equalExceptExtra(a, b []Sam) bool {
+	if len(a) != len(b) {
+		return false
+	}
+	for i := range a {
+		if a[i].QName != b[i].QName {
+			return false
+		}
+		if a[i].Flag != b[i].Flag {
+			return false
+		}
+		if a[i].RName != b[i].RName {
+			return false
+		}
+		if a[i].Pos != b[i].Pos {
+			return false
+		}
+		if a[i].MapQ != b[i].MapQ {
+			return false
+		}
+		if cigar.ToString(a[i].Cigar) != cigar.ToString(b[i].Cigar) {
+			return false
+		}
+		if a[i].RNext != b[i].RNext {
+			return false
+		}
+		if a[i].PNext != b[i].PNext {
+			return false
+		}
+		if a[i].TLen != b[i].TLen {
+			return false
+		}
+		if dna.CompareSeqsIgnoreCase(a[i].Seq, b[i].Seq) != 0 {
+			return false
+		}
+		if a[i].Qual != b[i].Qual {
+			return false
+		}
+	}
+	return true
+}
+
+const bigBam string = "/Users/danielsnellings/Desktop/10k.bam"
+
+func BenchmarkBamOpenClose(b *testing.B) {
+	for i := 0; i < b.N; i++ {
+		r, _ := OpenBam(bigBam)
+		r.Close()
+	}
+}
+
+func BenchmarkBamAllocs(b *testing.B) {
+	r, _ := OpenBam(bigBam)
+	var s Sam
+	for i := 0; i < b.N; i++ {
+		_, _ = DecodeBam(r, &s)
+		s.RName = ""
+		b.StopTimer()
+		if i%9000 == 0 {
+			r.Close()
+			r, _ = OpenBam(bigBam)
+		}
+		b.StartTimer()
+	}
+	r.Close()
+}
+
+func BenchmarkGonomicsBamRead(b *testing.B) {
+	for i := 0; i < b.N; i++ {
+		r, _ := OpenBam(bigBam)
+		var s Sam
+		var err error
+		for {
+			_, err = DecodeBam(r, &s)
+			if err == io.EOF {
+				break
+			}
+			fmt.Fprint(ioutil.Discard, s)
+		}
+		r.Close()
+	}
+}
+
+func BenchmarkBiogoBamRead(b *testing.B) {
```

Comment:
Another cool bam reader I found is from grail: [https://github.com/grailbio/bio/tree/master/encoding/bam](https://github.com/grailbio/bio/tree/master/encoding/bam). I'd be interested to see how your benchmarks compare to this implementation

Reply:
Nice! I didn't know Grail had one. I ran a couple benchmarks to check it out (grail func below).


For a printing task they they perform pretty comparably in terms of speed (gonomics looks to be marginally faster), but grail has ~50% fewer allocs/op
```
BenchmarkGonomicsBamRead-8            10         101756525 ns/op        37637304 B/op    1115598 allocs/op
BenchmarkGrailBamRead-8               10         107098446 ns/op        33841868 B/op     616302 allocs/op
```

For a straight reading task (just removed `fmt.Fprintf(ioutil.Discard, read)`), grail wins by a pretty substantial margin on time. Gonomics has slightly fewer allocs/op, but uses significantly less memory (2.6MB vs 15.2MB).

```
BenchmarkGonomicsBamRead-8            58          20298426 ns/op         2623551 B/op      33174 allocs/op
BenchmarkGrailBamRead-8               82          14090170 ns/op        15241928 B/op      36918 allocs/op
```

One potential saving grace is that on my machine, the grail reader uses 8 threads for reading whereas the gonomics reader currently uses only 1 thread. I will need to add a concurrent reader and see if we get any performance bump. 

Not familiar with how to operate this package so I copied a func from one of their benchmarks. Let me know if you think this should be change to make it a better test.
```
func BenchmarkGrailBamRead(b *testing.B) {
	for i := 0; i < b.N; i++ {
		provider := bamprovider.NewProvider(bigBam)
		iters, err := bamprovider.NewPairIterators(provider, true)
		if err != nil {
			b.Error(err)
		}
		var waitGroup sync.WaitGroup
		waitGroup.Add(len(iters))
		for i, iter := range iters {
			go func(id int, iter *bamprovider.PairIterator) {
				for iter.Scan() {
					record := iter.Record()
					fmt.Fprint(ioutil.Discard, record.R1)
					fmt.Fprint(ioutil.Discard, record.R2)
				}
				waitGroup.Done()
			}(i, iter)
		}
		waitGroup.Wait()
		if err := provider.Close(); err != nil {
			b.Error(err)
		}
	}
}
```

Reply:
Awesome, I'll take a closer look later today. I forget what method/function it is, but I know there's a way to set a cap to number of threads used during the benchmarks.